score/deployment/deployment.go

```diff
@@ -0,0 +1,108 @@
+			if internal.LabelSelectorMatchesLabels(svcSelector, deployment.Spec.Template.Labels) {
+				referencedByService = true
+				break
+			}
+		}
+
+		if referencedByService {
+			if deployment.Spec.Strategy.Type == v1.RollingUpdateDeploymentStrategyType {
+				score.Grade = scorecard.GradeAllOK
+			} else {
+				score.Grade = scorecard.GradeWarning
+				score.AddCommentWithURL("", "Deployment update strategy", "The deployment is used by a service but not using rolling update strategy which can cause interruptions", "https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#strategy")
+			}
+		} else {
+			score.Skipped = true
+			score.AddComment("", "Skipped as the deployment strategy does not matter if not targeted by a service", "")
+		}
+
+		return
+	}
+}
+
+// deploymentReplicas checks if a Deployment has >= 2 replicas if not (targeted by service || has HorizontalPodAutoscaler)
+func deploymentReplicas(svcs []ks.Service, hpas []ks.HpaTargeter) func(deployment v1.Deployment) (scorecard.TestScore, error) {
+	svcsInNamespace := make(map[string][]map[string]string)
+	for _, s := range svcs {
+		svc := s.Service()
+		if _, ok := svcsInNamespace[svc.Namespace]; !ok {
+			svcsInNamespace[svc.Namespace] = []map[string]string{}
+		}
+		svcsInNamespace[svc.Namespace] = append(svcsInNamespace[svc.Namespace], svc.Spec.Selector)
+	}
+	hpasInNamespace := make(map[string][]autoscalingv1.CrossVersionObjectReference)
+	for _, hpa := range hpas {
+		hpaTarget := hpa.HpaTarget()
+		hpaMeta := hpa.GetObjectMeta()
+		if _, ok := hpasInNamespace[hpaMeta.Namespace]; !ok {
+			hpasInNamespace[hpaMeta.Namespace] = []autoscalingv1.CrossVersionObjectReference{}
+		}
+		hpasInNamespace[hpaMeta.Namespace] = append(hpasInNamespace[hpaMeta.Namespace], hpaTarget)
+	}
+
+	return func(deployment v1.Deployment) (score scorecard.TestScore, err error) {
+		referencedByService := false
+		hasHPA := false
+
+		for _, svcSelector := range svcsInNamespace[deployment.Namespace] {
+			if internal.LabelSelectorMatchesLabels(svcSelector, deployment.Spec.Template.Labels) {
+				referencedByService = true
+				break
+			}
+		}
+
+		for _, hpaTarget := range hpasInNamespace[deployment.Namespace] {
+			if deployment.TypeMeta.APIVersion == hpaTarget.APIVersion &&
+				deployment.TypeMeta.Kind == hpaTarget.Kind &&
+				deployment.ObjectMeta.Name == hpaTarget.Name {
+				hasHPA = true
+				break
+			}
+		}
+
+		if !referencedByService || hasHPA {
+			score.Skipped = true
+			score.AddComment("", "Skipped as the deployment replicas does not matter if not targeted by service or if managed by HorizontalPodAutoscaler", "")
+		} else {
+			if ptr.Deref(deployment.Spec.Replicas, 1) >= 2 {
+				score.Grade = scorecard.GradeAllOK
+			} else {
+				score.Grade = scorecard.GradeWarning
+				score.AddComment("", "Deployment replica set", "The deployment is used by a service, not managed by HorizontalPodAutoscaler and has less then 2 replicas")
```

Comment:
```suggestion
				score.AddComment("", "Deployment few replicas", "Deployments targeted by Services are recommended to have at least 2 replicas to prevent unwanted downtime.")
```